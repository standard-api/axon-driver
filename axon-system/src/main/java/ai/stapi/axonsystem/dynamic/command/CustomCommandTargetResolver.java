package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
