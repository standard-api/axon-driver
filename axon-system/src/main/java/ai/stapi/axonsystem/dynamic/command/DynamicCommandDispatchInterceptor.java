package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
