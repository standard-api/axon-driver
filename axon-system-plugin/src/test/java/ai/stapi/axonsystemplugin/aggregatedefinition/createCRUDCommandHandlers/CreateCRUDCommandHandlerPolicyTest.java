package ai.stapi.axonsystemplugin.aggregatedefinition.createCRUDCommandHandlers;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.domain.DomainTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.springframework.test.context.ActiveProfiles;

@StructureDefinitionScope(CreateCrudCommandHandlerPolicyTestDefinitionsLoader.SCOPE)
@ActiveProfiles("dev")
class CreateCRUDCommandHandlerPolicyTest extends DomainTestCase {

  @Test
  void itCanCreateBasicCrudOperationWhenAggregateDefinitionAdded() {
    var expectedAggregateIdentifier = new UniqueIdentifier("ExampleTestResourceAggregate");
    this.whenCommandIsDispatched(
        new DynamicCommand(
            expectedAggregateIdentifier,
            "CreateAggregateDefinition",
            Map.of(
                "id", "ExampleTestResourceAggregate",
                "name", "ExampleTestResource",
                "description", "This is a example test aggregate.",
                "tag", new ArrayList<>(List.of("tag1", "tag2", "tag3")),
                "structure", new HashMap<>(Map.of(
                    "id", "ExampleTestResource"
                ))
            )
        )
    );
    this.thenExpectedDynamicEventsSaved(
        "AggregateDefinitionCreated",
        "OperationDefinitionCreated",
        "EventMessageDefinitionCreated",
        "CommandOnAggregateDefinitionAdded",
        "OperationDefinitionCreated",
        "EventMessageDefinitionCreated",
        "CommandOnAggregateDefinitionAdded",
        "OperationDefinitionCreated",
        "EventMessageDefinitionCreated",
        "CommandOnAggregateDefinitionAdded"
    );
    this.thenMergedGraphOfAggregateApproved(expectedAggregateIdentifier);
  }
}
