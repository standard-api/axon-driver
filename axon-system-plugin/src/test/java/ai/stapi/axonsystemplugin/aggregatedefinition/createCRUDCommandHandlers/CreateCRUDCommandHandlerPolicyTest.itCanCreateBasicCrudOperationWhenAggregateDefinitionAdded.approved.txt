nodes:
	node_type: AggregateDefinition
	node_hash: 4CA35
	node_edges:
		AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 5EFB4 -> EC773)
		AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 9A2C1 -> D1A1B)
		AggregateDefinition -> structure -> StructureDefinition (4CA35 -> 1D20A -> DABF6)
	attributes:
		description -> This is a example test aggregate.
		name -> ExampleTestResource

	node_type: CommandHandlerDefinition
	node_hash: D1A1B
	node_edges:
		AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 9A2C1 -> D1A1B)
		CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (D1A1B -> 54A69 -> 5C13E)
		CommandHandlerDefinition -> operation -> OperationDefinition (D1A1B -> E728B -> 8103B)
	attributes:
		creationalPolicy -> if_missing

	node_type: CommandHandlerDefinition
	node_hash: EC773
	node_edges:
		AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 5EFB4 -> EC773)
		CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (EC773 -> 43A0B -> 51932)
		CommandHandlerDefinition -> operation -> OperationDefinition (EC773 -> DA318 -> 4DE01)
	attributes:
		creationalPolicy -> never

	node_type: CommandHandlerDefinitionEventFactory
	node_hash: 51932
	node_edges:
		CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (EC773 -> 43A0B -> 51932)
		CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (51932 -> FE71B -> BBC49)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (51932 -> 58078 -> 9E2DE)
	attributes:

	node_type: CommandHandlerDefinitionEventFactory
	node_hash: 5C13E
	node_edges:
		CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (D1A1B -> 54A69 -> 5C13E)
		CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (5C13E -> FE71B -> BBC49)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 056BF -> BC65A)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 16A40 -> 95423)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 37A4E -> 45724)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4144C -> B029A)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 49412 -> A5196)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4B593 -> 98EC2)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 58078 -> 9E2DE)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 6D621 -> F0318)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 989F7 -> BD30E)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 9FDD8 -> 1DDC6)
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> B86E8 -> F6DF5)
	attributes:

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 1DDC6
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 9FDD8 -> 1DDC6)
	attributes:
		inputValueParameterName -> language
		kind -> add
		modificationPath -> language

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 45724
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 37A4E -> 45724)
	attributes:
		inputValueParameterName -> modifierExtension
		kind -> add
		modificationPath -> modifierExtension

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 95423
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 16A40 -> 95423)
	attributes:
		inputValueParameterName -> extension
		kind -> add
		modificationPath -> extension

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 98EC2
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4B593 -> 98EC2)
	attributes:
		inputValueParameterName -> implicitRules
		kind -> add
		modificationPath -> implicitRules

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 9E2DE
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (51932 -> 58078 -> 9E2DE)
	attributes:
		inputValueParameterName -> tag
		kind -> add
		modificationPath -> tag

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: 9E2DE
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 58078 -> 9E2DE)
	attributes:
		inputValueParameterName -> tag
		kind -> add
		modificationPath -> tag

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: A5196
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 49412 -> A5196)
	attributes:
		inputValueParameterName -> name
		kind -> add
		modificationPath -> name

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: B029A
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4144C -> B029A)
	attributes:
		inputValueParameterName -> structure
		kind -> add
		modificationPath -> structure

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: BC65A
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 056BF -> BC65A)
	attributes:
		inputValueParameterName -> contained
		kind -> add
		modificationPath -> contained

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: BD30E
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 989F7 -> BD30E)
	attributes:
		inputValueParameterName -> description
		kind -> add
		modificationPath -> description

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: F0318
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 6D621 -> F0318)
	attributes:
		inputValueParameterName -> text
		kind -> add
		modificationPath -> text

	node_type: CommandHandlerDefinitionEventFactoryModification
	node_hash: F6DF5
	node_edges:
		CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> B86E8 -> F6DF5)
	attributes:
		inputValueParameterName -> meta
		kind -> add
		modificationPath -> meta

	node_type: EventMessageDefinition
	node_hash: BBC49
	node_edges:
		CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (51932 -> FE71B -> BBC49)
	attributes:

	node_type: EventMessageDefinition
	node_hash: BBC49
	node_edges:
		CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (5C13E -> FE71B -> BBC49)
	attributes:

	node_type: OperationDefinition
	node_hash: 4DE01
	node_edges:
		CommandHandlerDefinition -> operation -> OperationDefinition (EC773 -> DA318 -> 4DE01)
	attributes:

	node_type: OperationDefinition
	node_hash: 8103B
	node_edges:
		CommandHandlerDefinition -> operation -> OperationDefinition (D1A1B -> E728B -> 8103B)
	attributes:

	node_type: StructureDefinition
	node_hash: DABF6
	node_edges:
		AggregateDefinition -> structure -> StructureDefinition (4CA35 -> 1D20A -> DABF6)
	attributes:

edges:
	edge_type: command
	edge_hash: 5EFB4
	AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 5EFB4 -> EC773)
	attributes:

	edge_type: command
	edge_hash: 9A2C1
	AggregateDefinition -> command -> CommandHandlerDefinition (4CA35 -> 9A2C1 -> D1A1B)
	attributes:

	edge_type: event
	edge_hash: FE71B
	CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (51932 -> FE71B -> BBC49)
	attributes:

	edge_type: event
	edge_hash: FE71B
	CommandHandlerDefinitionEventFactory -> event -> EventMessageDefinition (5C13E -> FE71B -> BBC49)
	attributes:

	edge_type: eventFactory
	edge_hash: 43A0B
	CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (EC773 -> 43A0B -> 51932)
	attributes:

	edge_type: eventFactory
	edge_hash: 54A69
	CommandHandlerDefinition -> eventFactory -> CommandHandlerDefinitionEventFactory (D1A1B -> 54A69 -> 5C13E)
	attributes:

	edge_type: modification
	edge_hash: 056BF
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 056BF -> BC65A)
	attributes:

	edge_type: modification
	edge_hash: 16A40
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 16A40 -> 95423)
	attributes:

	edge_type: modification
	edge_hash: 37A4E
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 37A4E -> 45724)
	attributes:

	edge_type: modification
	edge_hash: 4144C
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4144C -> B029A)
	attributes:

	edge_type: modification
	edge_hash: 49412
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 49412 -> A5196)
	attributes:

	edge_type: modification
	edge_hash: 4B593
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 4B593 -> 98EC2)
	attributes:

	edge_type: modification
	edge_hash: 58078
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (51932 -> 58078 -> 9E2DE)
	attributes:

	edge_type: modification
	edge_hash: 58078
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 58078 -> 9E2DE)
	attributes:

	edge_type: modification
	edge_hash: 6D621
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 6D621 -> F0318)
	attributes:

	edge_type: modification
	edge_hash: 989F7
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 989F7 -> BD30E)
	attributes:

	edge_type: modification
	edge_hash: 9FDD8
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> 9FDD8 -> 1DDC6)
	attributes:

	edge_type: modification
	edge_hash: B86E8
	CommandHandlerDefinitionEventFactory -> modification -> CommandHandlerDefinitionEventFactoryModification (5C13E -> B86E8 -> F6DF5)
	attributes:

	edge_type: operation
	edge_hash: DA318
	CommandHandlerDefinition -> operation -> OperationDefinition (EC773 -> DA318 -> 4DE01)
	attributes:

	edge_type: operation
	edge_hash: E728B
	CommandHandlerDefinition -> operation -> OperationDefinition (D1A1B -> E728B -> 8103B)
	attributes:

	edge_type: structure
	edge_hash: 1D20A
	AggregateDefinition -> structure -> StructureDefinition (4CA35 -> 1D20A -> DABF6)
	attributes:
